cli/src/commands/cloud/environments/tunnel/mod.rs

```diff
@@ -0,0 +1,190 @@
+        _ => return Err(anyhow::anyhow!("First switch to a Cloud environment using `restate config use-environment` or configure one with `restate cloud environment configure`"))
+    };
+
+    let bearer_token = if let Some(bearer_token) = &env.config.bearer_token {
+        // the user may have specifically set an api token
+        bearer_token.clone()
+    } else if let Some(cloud_credentials) = &env.config.cloud.credentials {
+        cloud_credentials.access_token()?.to_string()
+    } else {
+        return Err(anyhow::anyhow!(
+            "Restate Cloud credentials have not been provided; first run `restate cloud login`"
+        ));
+    };
+
+    let mut opts = opts.clone();
+    if opts.local_port.is_none() && opts.remote_port.is_empty() {
+        opts.local_port = Some(9080);
+        opts.remote_port = vec![RemotePort::Ingress, RemotePort::Admin];
+    }
+
+    let client = CloudClient::new(&env)?;
+
+    let environment_info = client
+        .describe_environment(
+            &environment_info.account_id,
+            &environment_info.environment_id,
+        )
+        .await?
+        .into_body()
+        .await?;
+
+    let client = reqwest::Client::builder()
+        .user_agent(format!(
+            "{}/{} {}-{}",
+            env!("CARGO_PKG_NAME"),
+            build_info::RESTATE_CLI_VERSION,
+            std::env::consts::OS,
+            std::env::consts::ARCH,
+        ))
+        .connect_timeout(env.connect_timeout)
+        .http2_prior_knowledge()
+        .build()?;
+
+    let cancellation = CancellationToken::new();
+
+    // prevent ctrl-c from clearing the screen, instead just cancel
+    {
+        let cancellation = cancellation.clone();
+        let boxed: Box<dyn Fn() + Send> = Box::new(move || cancellation.cancel());
+        *crate::EXIT_HANDLER.lock().unwrap() = Some(boxed);
+    }
+
+    let tunnel_renderer = Arc::new(renderer::TunnelRenderer::new(&opts.remote_port).unwrap());
+
+    let local_fut = local::run_local(
+        &env,
+        client.clone(),
+        &bearer_token,
+        environment_info,
+        &opts,
+        tunnel_renderer.clone(),
+    );
+
+    let remote_futs = futures::stream::FuturesUnordered::new();
+    for remote_port in &opts.remote_port {
+        let base_url = match remote_port {
+            remote::RemotePort::Ingress => env.ingress_base_url()?,
+            remote::RemotePort::Admin => env.admin_base_url()?,
+        };
+
+        remote_futs.push(remote::run_remote(
+            *remote_port,
+            client.clone(),
+            base_url,
+            &bearer_token,
+            tunnel_renderer.clone(),
+        ));
+    }
+
+    let mut rerender = tokio::time::interval(Duration::from_millis(100));
+
+    let res = {
+        tokio::pin!(local_fut);
+        let mut remote_futs = remote_futs;
+
+        loop {
+            tokio::select! {
+                Err(local_err) = &mut local_fut => break Err(Error::Local(local_err)),
```

Comment:
The panic on setting a remote-port but no local-port goes away if I change this line to:

```suggestion
                Err(local_err) = &mut local_fut, if opts.local_port.is_some() => break Err(Error::Local(local_err)),
```

Though I'm not sure why it works that way, since `local_fut` should evaluate to `Ok(())`  when local port is unset?

Reply:
Need to fuse the future otherwise it will panic when polled again after already returning Poll::Ready(Ok(())). Bit of a footgun with tokio::select!.
https://users.rust-lang.org/t/why-doesnt-tokio-select-require-fusedfuture/46975/13